### PR TITLE
Use full requirement when serializing receipt

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

The current receipt doesn't capture quite enough information. For example, it doesn't differentiate between editable and non-editable requirements. This PR instead uses the full `Requirement` type. I think we should use a custom representation like we do in the lockfile, but I'm just using the default representation to demonstrate the idea.

### CODE REVIEW

File: crates/uv/tests/tool_install.rs [hunk 1]
```diff
@@ -396,9 +396,7 @@ fn tool_install_editable() {
     ----- stderr -----
     "###);
 
-    // Request `black`. It should retain the current installation.
-    // TODO(charlie): This is arguably incorrect, especially because the tool receipt removes the
-    // file path.
```
Comment (charliermarsh) [inline crates/uv/tests/tool_install.rs:401]: This was the motivating example.

File: crates/uv/tests/tool_install.rs [hunk 2]
```diff
@@ -380,7 +380,7 @@ fn tool_install_editable() {
         // We should have a tool receipt
         assert_snapshot!(fs_err::read_to_string(tool_dir.join("black").join("uv-receipt.toml")).unwrap(), @r###"
         [tool]
-        requirements = ["black @ file://[WORKSPACE]/scripts/packages/black_editable"]
+        requirements = [{ name = "black", directory = { install_path = "[WORKSPACE]/scripts/packages/black_editable", lock_path = "[WORKSPACE]/scripts/packages/black_editable", editable = true, url = { url = "file://[WORKSPACE]/scripts/packages/black_editable" } } }]
```
Comment (charliermarsh) [inline crates/uv/tests/tool_install.rs:383]: If we're comfortable with the general idea here, I will iron out the actual requirement schema.

**Comment by charliermarsh** [comment]
Please just focus on the concept here ([example](https://github.com/astral-sh/uv/pull/5494/files#r1693648630)) and ignore the code and exact representation for now.

**Comment by zanieb** [comment]
I think this makes sense to me. My only concern is that I want users to be able to declare tools declaratively eventually. I don't think that's in conflict though — I think the receipt can be considered "machine read / write" and "human inspectable" like the lock file. We can iron out a separate schema for users to declaratively request tools later.

**Comment by BurntSushi** [comment]
I'm not terribly familiar with tool receipts yet. Can you say more at a high level about what problem this is trying to solve?

(Insomuch as I understand what this PR is doing, I think it LGTM.)

**Comment by charliermarsh** [comment]
Copying from Discord:

```
Yes sorry
When a user installs a tool (like cargo install), we save a receipt alongside it enumerating the input requirements
When they run uv tool upgrade, we respect those input requirements
...
It was using PEP 508 so...
If a user did tool install -e ./black, we didn't capture that it was installed as editable, for example
Since that can't be expressed in PEP 508
But our requirement type is richer
```

**Comment by charliermarsh** [comment]
Ok happy with how this turned out.

**Comment by charliermarsh** [comment]
Ready for review @zanieb @BurntSushi at your convenience.

File: crates/pypi-types/src/requirement.rs [hunk 3]
```diff
@@ -415,3 +432,229 @@ impl RequirementSource {
...
+                Ok(RequirementSource::Directory {
+                    url: VerbatimUrl::from_path(editable.as_path())?,
+                    install_path: editable.clone(),
+                    lock_path: editable,
+                    editable: true,
+                })
+            }
+        }
+    }
+}
```
Comment (BurntSushi) [inline crates/pypi-types/src/requirement.rs:660]: If we're using a Git URL here, perhaps we should in the lock file too? Then I could just close https://github.com/astral-sh/uv/pull/4631.

**Review by BurntSushi — APPROVED** [review]
LGTM

File: crates/pypi-types/src/requirement.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/pypi-types/src/requirement.rs:660]: That seems ok to me.

Comment (charliermarsh) [reply crates/pypi-types/src/requirement.rs:660]: Let me re-review that quickly, I think I lost it.

Comment (charliermarsh) [reply crates/pypi-types/src/requirement.rs:660]: Honestly, I'm fine with either!

**Pull request merged**